File: Trunk/vpPriV100GrupoMundifios/SugereLotes/Compras/EditorCompras/CmpIsEditorCompras.cs
using Generico;
using Microsoft.VisualBasic;
using Primavera.Extensibility.BusinessEntities.ExtensibilityService.EventArgs;
using Primavera.Extensibility.Purchases.Editors;
using StdBE100;
using System;

namespace SugereLotes
{
    public class CmpIsEditorCompras : EditorCompras
    {
        public override void ArtigoIdentificado(string Artigo, int NumLinha, ref bool Cancel, ExtensibilityEventArgs e)
            {
            base.ArtigoIdentificado(Artigo, NumLinha, ref Cancel, e);

            if (Module1.VerificaToken("SugereLotes") == 1)
            {
                if (DocumentoCompra.Entidade + "" != "" && BSO.Compras.TabCompras.Edita(DocumentoCompra.Tipodoc).TipoDocumento == 2 && (BSO.Base.Artigos.Edita(Artigo).Descricao.StartsWith("Fio") || BSO.Base.Artigos.Edita(Artigo).Descricao.StartsWith("Rama")))
                {
                    // Sugestão de lote
                    int i;
                    string ent;
                    int lote;
                    int loteAux;
                    StdBELista listLote;

                    ent = this.DocumentoCompra.Entidade;
                    loteAux = 0;
                    // Consulta à função dbo.fnProximoLote de qual o proximo lote a utilizar.
                    listLote = BSO.Consulta("select PRIMUNDIFIOS.dbo.fnProximoLote('" + BSO.Base.Fornecedores.Edita(this.DocumentoCompra.Entidade).CamposUtil["CDU_EntidadeInterna"].Valor + "','" + Artigo + "') as 'Lote'");

                    listLote.Inicio();
                    for (i = 1; i <= DocumentoCompra.Linhas.NumItens; i++)
                    {
                        if (DocumentoCompra.Linhas.GetEdita(i).Artigo == Artigo && i != NumLinha && DocumentoCompra.Linhas.GetEdita(i).Lote.Length == 8)
                        {
                            lote = Convert.ToInt32(Strings.Right(this.DocumentoCompra.Linhas.GetEdita(i).Lote, 4));

                            if (lote > loteAux)
                                loteAux = lote;
                        }
                    }

                    if (loteAux != 0)
                    {
                        loteAux++;

                        if (listLote.Vazia() || listLote.Valor("Lote") == string.Empty)
                        {
                            i = 4 - Strings.Len(Convert.ToString(loteAux));
                            this.DocumentoCompra.Linhas.GetEdita(NumLinha).Lote = "" + ent + Strings.Left("0000", i) + loteAux;
                        }
                        else
                        {
                            if (loteAux <= Convert.ToInt32(Strings.Right(listLote.Valor("Lote"), 4)))
                                this.DocumentoCompra.Linhas.GetEdita(NumLinha).Lote = listLote.Valor("Lote");
                            else
                            {
                                i = 4 - Strings.Len(Convert.ToString(loteAux));
                                this.DocumentoCompra.Linhas.GetEdita(NumLinha).Lote = "" + ent + Strings.Left("0000", i) + loteAux;
                            }
                        }
                    }
                    else
                    {
                        if (listLote.Vazia() || listLote.Valor("Lote") == string.Empty)
                            DocumentoCompra.Linhas.GetEdita(NumLinha).Lote = "" + ent + "0001";
                        else
                            DocumentoCompra.Linhas.GetEdita(NumLinha).Lote = listLote.Valor("Lote");
                    }
                }
            }
        }
    }
}

File: Branches/VBInicial/vpPriV100GrupoMundifios/SugereLotes/Compras/EditorCompras/CmpIsEditorCompras.vb
Imports Primavera.Extensibility.BusinessEntities.ExtensibilityService.EventArgs
Imports Primavera.Extensibility.Purchases.Editors
Imports Primavera.Extensibility.Attributes
Imports Vimaponto.PrimaveraV100.Clientes.GrupoMundifios.Generico
Imports StdBE100

Namespace SugereLotes
    Public Class CmpIsEditorCompras
        Inherits EditorCompras

        Public Overrides Sub ArtigoIdentificado(Artigo As String, NumLinha As Integer, ByRef Cancel As Boolean, e As ExtensibilityEventArgs)
            MyBase.ArtigoIdentificado(Artigo, NumLinha, Cancel, e)


            If Module1.VerificaToken("SugereLotes") = 1 Then

                '############################################################################################
                '####              JFC 21/10/2019 Sugestão de Lotes                            ##############
                '############################################################################################

                If Me.DocumentoCompra.Entidade & "" <> "" And BSO.Compras.TabCompras.Edita(Me.DocumentoCompra.Tipodoc).TipoDocumento = 2 And (BSO.Base.Artigos.Edita(Artigo).Descricao Like "Fio*" Or BSO.Base.Artigos.Edita(Artigo).Descricao Like "Rama*") Then
                    'Sugestão de lote
                    Dim i As Long
                    Dim ent As String
                    Dim lote As Integer
                    Dim loteAux As Integer
                    Dim listLote As StdBELista

                    ent = Me.DocumentoCompra.Entidade
                    loteAux = 0
                    'Consulta à função dbo.fnProximoLote de qual o proximo lote a utilizar.
                    listLote = BSO.Consulta("select PRIMUNDIFIOS.dbo.fnProximoLote('" & BSO.Base.Fornecedores.Edita(Me.DocumentoCompra.Entidade).CamposUtil("CDU_EntidadeInterna").Valor & "','" & Artigo & "') as 'Lote'")

                    listLote.Inicio()

                    'Primeira validação, ver se já existe nas outras linhas algum lote já inserido
                    For i = 1 To Me.DocumentoCompra.Linhas.NumItens

                        If Me.DocumentoCompra.Linhas.GetEdita(i).Artigo = Artigo And i <> NumLinha And Len(Me.DocumentoCompra.Linhas.GetEdita(i).Lote) = 8 Then

                            lote = CInt(Right(Me.DocumentoCompra.Linhas.GetEdita(i).Lote, 4))
                            If lote > loteAux Then
                                loteAux = lote
                            End If

                        End If
                    Next i

                    'Se já existir o lote noutra linha, então lote nove = lote +1
                    If loteAux <> 0 Then
                        loteAux = loteAux + 1

                        'Validar se o lote novo é igual ou superior ao lote sugerido pela funcção dbo.fnProximoLote
                        If listLote.Vazia Or listLote.Valor("Lote") = "" Then
                            i = 4 - Len(CStr(loteAux))
                            Me.DocumentoCompra.Linhas.GetEdita(NumLinha).Lote = "" & ent & Left("0000", i) & loteAux
                        Else
                            If loteAux <= CInt(Right(listLote.Valor("Lote"), 4)) Then
                                Me.DocumentoCompra.Linhas.GetEdita(NumLinha).Lote = listLote.Valor("Lote")
                            Else
                                i = 4 - Len(CStr(loteAux))
                                Me.DocumentoCompra.Linhas.GetEdita(NumLinha).Lote = "" & ent & Left("0000", i) & loteAux
                            End If
                        End If

                        'Não foi detectado nenhum lote inserido
                    Else

                        'Se não existir nenhum lote sugerido pela funcção dbo.fnProximoLote, então é o primeiro lote
                        If listLote.Vazia Or listLote.Valor("Lote") = "" Then
                            Me.DocumentoCompra.Linhas.GetEdita(NumLinha).Lote = "" & ent & "0001"
                            'Lote sugerido pela função
                        Else
                            Me.DocumentoCompra.Linhas.GetEdita(NumLinha).Lote = listLote.Valor("Lote")
                        End If
                    End If
                End If
                '############################################################################################
                '####              JFC 21/10/2019 Sugestão de Lotes                            ##############
                '############################################################################################


            End If

        End Sub


    End Class
End Namespace
